experimental/devices/mfrc522/mfrc522.go

```diff
@@ -0,0 +1,757 @@
+	}
+	if irqPin == nil {
+		return nil, wrapf("IRQ pin is not set")
+	}
+
+	spiDev, err := spiPort.Connect(10000000, spi.Mode0, 8)
+	if err != nil {
+		return nil, err
+	}
+
+	dev := &Dev{
+		spiDev:           spiDev,
+		operationTimeout: 30 * time.Second,
+	}
+
+	if err := resetPin.Out(gpio.High); err != nil {
+		return nil, err
+	}
+
+	dev.resetPin = resetPin
+
+	if err := irqPin.In(gpio.PullUp, gpio.FallingEdge); err != nil {
+		return nil, err
+	}
+
+	dev.irqPin = irqPin
+
+	if err := dev.Init(); err != nil {
+		return nil, err
+	}
+
+	return dev, nil
+}
+
+// SetOperationtimeout updates the device timeout for card operations.
+// Effectively that sets the maximum time the RFID device will wait for IRQ from the proximity card detection.
+//
+//	timeout the duration to wait for IRQ strobe.
+func (r *Dev) SetOperationtimeout(timeout time.Duration) {
+	r.operationTimeout = timeout
+}
+
+// Init initializes the RFID chip.
+func (r *Dev) Init() error {
+	if err := r.Reset(); err != nil {
+		return err
+	}
+	if err := r.writeCommandSequence(sequenceCommands.init); err != nil {
+		return err
+	}
+	return r.SetAntenna(true)
+}
+
+// Reset resets the RFID chip to initial state.
+func (r *Dev) Reset() error {
+	return r.devWrite(commands.CommandReg, commands.PCD_RESETPHASE)
+}
+
+// Halt soft-stops the chip - PowerDown bit set, command IDLE
+func (r *Dev) Halt() error {
+	return r.devWrite(commands.CommandReg, 16)
+}
+
+// SetAntenna configures the antenna state, on/off.
+func (r *Dev) SetAntenna(state bool) error {
+	if state {
+		current, err := r.devRead(commands.TxControlReg)
+		if err != nil {
+			return err
+		}
+		if current&0x03 != 0 {
+			return wrapf("can not set the bitmask for antenna")
+		}
+		return r.setBitmask(commands.TxControlReg, 0x03)
+	}
+	return r.clearBitmask(commands.TxControlReg, 0x03)
+}
+
+// CardWrite the low-level interface to write some raw commands to the card.
+//
+// 	command - the command register
+// 	data - the data to write out to the card using the authenticated sector.
+func (r *Dev) CardWrite(command byte, data []byte) ([]byte, int, error) {
+	var backData []byte
+	backLength := -1
+	irqEn := byte(0x00)
+	irqWait := byte(0x00)
+
+	switch command {
+	case commands.PCD_AUTHENT:
+		irqEn = 0x12
+		irqWait = 0x10
+	case commands.PCD_TRANSCEIVE:
+		irqEn = 0x77
+		irqWait = 0x30
+	}
+
+	r.devWrite(commands.CommIEnReg, irqEn|0x80)
+	r.clearBitmask(commands.CommIrqReg, 0x80)
+	r.setBitmask(commands.FIFOLevelReg, 0x80)
+	r.devWrite(commands.CommandReg, commands.PCD_IDLE)
+
+	for _, v := range data {
+		r.devWrite(commands.FIFODataReg, v)
+	}
+
+	r.devWrite(commands.CommandReg, command)
+
+	if command == commands.PCD_TRANSCEIVE {
+		r.setBitmask(commands.BitFramingReg, 0x80)
+	}
+
+	i := 2000
+	n := byte(0)
+
+	for ; i > 0; i-- {
+		n, err := r.devRead(commands.CommIrqReg)
+		if err != nil {
+			return nil, -1, err
+		}
+		if n&(irqWait|1) != 0 {
+			break
+		}
+	}
+
+	r.clearBitmask(commands.BitFramingReg, 0x80)
```

Comment:
still here

I know it's tedious, but after 10+ years of python, I still prefer to check manually than to have to catch random exceptions.